<a href="https://colab.research.google.com/github/M-Dogra/Python/blob/main/Movie_Review_Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd


In [2]:
df = pd.read_csv('IMDB Dataset.csv')
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [3]:
# one review
df['review'][0]

"One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.<br /><br />The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word.<br /><br />It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many..Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.<br /><br />I would say the main appeal of the show is due to the fa

Data needs to be cleaned by removing html tags, special characters
#Steps:
1. Sample 10000 rows
2. Remove html tags
3. Remove special characters
4. Convert text to lower case
5. Remove Stop words: Stop words are common words in a language that do not add much meaning to text analysis. like "is," "the," "a," "an," "in," "at," "for," "of," "on," "and," "with," etc.
6. Stemming : text preprocessing technique used in Natural Language Processing (NLP), including sentiment analysis, to reduce words to their root or base form. e.g. "Running" → "Run"

In [4]:
df = df.sample(10000)
df.shape

(10000, 2)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10000 entries, 19730 to 31834
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     10000 non-null  object
 1   sentiment  10000 non-null  object
dtypes: object(2)
memory usage: 234.4+ KB


In [6]:
df['sentiment'].replace({'positive':1, 'negative':0}, inplace=True)

<ipython-input-6-8f30d573ea1c>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['sentiment'].replace({'positive':1, 'negative':0}, inplace=True)
<ipython-input-6-8f30d573ea1c>:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['sentiment'].replace({'positive':1, 'negative':0}, inplace=True)


In [7]:
df.head()

review  sentiment
19730  I don't know about you but i go to horror film...          0
6789   This movie was featured on a very early episod...          1
47303  After waking up at 3:30 in the morning and not...          0
46826  It really boggles my mind when someone comes a...          0
2018   I went to see this film out of curiosity, and ...          1

In [8]:
import re
clean = re.compile('<.*?>')
re.sub(clean, '', df.iloc[2].review)

'After waking up at 3:30 in the morning and not being able to go to sleep, I decided that I may as well be entertained while I suffered from sleep deprivation. This movie was on HBO and I decided to watch it. What a mistake!Here is the plot (potential spoilers, if you even care) : a neurotic man with an addiction to candy (Josh Kornbluth) works as a temp for some ridiculous company. Suddenly, he is offered the chance to "go perm," which seems to be the favorite catch-phrase of this movie. But with a secure job and a secure income within his grasp, Josh decides for whatever stupid reason that he doesn\'t want to do it. He hopelessly bungles several minor tasks and his laziness and lack of ambition fill me with contempt. His inability to send several "very important" letters within a given amount of time is supposed to be hilarious, but is merely stupid. Josh meets and deceives a lawyer woman and they wind up in bed together (obviously a fantasy of Josh\'s in real life, as such a thing w

In [9]:
#function to clean html tags
def clean_html(text):
    clean = re.compile('<.*?>')
    return re.sub(clean, '', text)

In [10]:
df['review']=df['review'].apply(clean_html)

In [11]:
df['review'].head()

19730    I don't know about you but i go to horror film...
6789     This movie was featured on a very early episod...
47303    After waking up at 3:30 in the morning and not...
46826    It really boggles my mind when someone comes a...
2018     I went to see this film out of curiosity, and ...
Name: review, dtype: object

In [12]:
# converting review to lower text

def convert_lower(text):
  return text.lower()

In [13]:
df['review'] = df['review'].apply(convert_lower)

In [14]:
# function to remove special characters

def remove_special(text):
  x=''

  for i in text:
    if i.isalnum():
      x=x+i
    else:
      x=x + ' '
  return x

In [15]:
df['review']=df['review'].apply(remove_special)

In [16]:
# Remove Stop words

import nltk
nltk.download('stopwords')
#natural language toolkit

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [17]:
from nltk.corpus import stopwords

In [18]:
print('Total number of stopwords in english =',len(stopwords.words('english')))

print(stopwords.words('english'))

Total number of stopwords in english = 179
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'n

In [19]:
df

review  sentiment
19730  i don t know about you but i go to horror film...          0
6789   this movie was featured on a very early episod...          1
47303  after waking up at 3 30 in the morning and not...          0
46826  it really boggles my mind when someone comes a...          0
2018   i went to see this film out of curiosity  and ...          1
...                                                  ...        ...
21093  i watched this movie last night and already i ...          0
8827   i finally got around to seeing this after hear...          1
751    they must issue this plot outline to all wanna...          0
5284   after having seen and loved postal  yes  i act...          0
31834  this is a movie that demonstrates that mood an...          0

[10000 rows x 2 columns]

In [20]:
def remove_stopwords(text):
  x=[]
  for i in text.split():

    if i not in stopwords.words('english'):
      x.append(i)
  y=x[:]
  x.clear()
  return y

In [21]:
df['review'] = df['review'].apply(remove_stopwords)

In [22]:
df

review  sentiment
19730  [know, go, horror, films, scared, anything, sc...          0
6789   [movie, featured, early, episode, mystery, sci...          1
47303  [waking, 3, 30, morning, able, go, sleep, deci...          0
46826  [really, boggles, mind, someone, comes, across...          0
2018   [went, see, film, curiosity, settle, argument,...          1
...                                                  ...        ...
21093  [watched, movie, last, night, already, struggl...          0
8827   [finally, got, around, seeing, hearing, great,...          1
751    [must, issue, plot, outline, wannabe, filmmake...          0
5284   [seen, loved, postal, yes, actually, loved, po...          0
31834  [movie, demonstrates, mood, music, texture, en...          0

[10000 rows x 2 columns]

In [23]:
# Stemming

from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [24]:
y=[]

def stem_words(text):
  for i in text:
    y.append(ps.stem(i))
  z=y[:]
  y.clear()
  return z

In [25]:
df['review']=df['review'].apply(stem_words)

In [26]:
df

review  sentiment
19730  [know, go, horror, film, scare, anyth, scari, ...          0
6789   [movi, featur, earli, episod, mysteri, scienc,...          1
47303  [wake, 3, 30, morn, abl, go, sleep, decid, may...          0
46826  [realli, boggl, mind, someon, come, across, mo...          0
2018   [went, see, film, curios, settl, argument, fil...          1
...                                                  ...        ...
21093  [watch, movi, last, night, alreadi, struggl, r...          0
8827   [final, got, around, see, hear, great, thing, ...          1
751    [must, issu, plot, outlin, wannab, filmmak, ar...          0
5284   [seen, love, postal, ye, actual, love, postal,...          0
31834  [movi, demonstr, mood, music, textur, enough, ...          0

[10000 rows x 2 columns]

In [27]:
# join back

def join_back(list_input):
  return " ".join(list_input)

In [28]:
df['review']=df['review'].apply(join_back)

In [29]:
df['review']

19730    know go horror film scare anyth scari movi sev...
6789     movi featur earli episod mysteri scienc theate...
47303    wake 3 30 morn abl go sleep decid may well ent...
46826    realli boggl mind someon come across movi like...
2018     went see film curios settl argument film best ...
                               ...                        
21093    watch movi last night alreadi struggl recollec...
8827     final got around see hear great thing actual e...
751      must issu plot outlin wannab filmmak arriv hol...
5284     seen love postal ye actual love postal decid t...
31834    movi demonstr mood music textur enough make go...
Name: review, Length: 10000, dtype: object

In [30]:
X = df.iloc[:,0:1].values

In [31]:
X.shape

(10000, 1)

In [32]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=1000)
#max_features=500

In [33]:
X= cv.fit_transform(df['review']).toarray()

In [34]:
X.shape

(10000, 1000)

In [35]:
y = df.iloc[:, -1].values

# these are the labels capturing only last column

In [36]:
y.shape

(10000,)

split the data into 2 parts
* 8000 rows : Training set
* 2000 rows : Test set
* Test set results are already known

In [37]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [38]:
X_train.shape

(8000, 1000)

In [39]:
X_test.shape

(2000, 1000)

In [40]:
y_train.shape

(8000,)

In [41]:
y_test.shape

(2000,)

In [42]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB

In [43]:
clf1 = GaussianNB()
clf2 = MultinomialNB()
clf3 = BernoulliNB()

In [44]:
clf1.fit(X_train, y_train)
clf2.fit(X_train, y_train)
clf3.fit(X_train, y_train)

BernoulliNB()

In [45]:
y_pred1=clf1.predict(X_test)
y_pred2=clf2.predict(X_test)
y_pred3=clf3.predict(X_test)

In [46]:
y_pred1.shape

(2000,)

In [47]:
#calculate accuracy

from sklearn.metrics import accuracy_score

In [48]:
print('Gaussian', accuracy_score(y_test, y_pred1))
print('Multinomial', accuracy_score(y_test, y_pred2))
print('Bernaulli', accuracy_score(y_test, y_pred3))

Gaussian 0.7975
Multinomial 0.8295
Bernaulli 0.8225


After performing sentiment analysis on a sample of 10000 from a dataset of 50,000 movie reviews, applying Naïve Bayes classifiers after text preprocessing. The accuracy scores with max_features=1000 are:
* Gaussian 0.79
* Multinomial 0.829
* Bernaulli 0.822

#Insights & Interpretation
1. Why is MultinomialNB the Best?

* ✅ Best for text classification – It assumes word frequency follows a multinomial distribution, making it ideal for Bag-of-Words (BoW) or TF-IDF representations.
* ✅ Captures word importance – Since sentiment is often determined by the presence of positive or negative words, MultinomialNB is well-suited for detecting sentiment patterns.

2. Why Did BernoulliNB Perform Similarly?
* ✅ BernoulliNB works well with binary features (word presence/absence).
* ✅ Since you removed stop words and limited vocabulary (max_features=1000), BernoulliNB was effective in sentiment classification.
* ✅ However, it ignores word frequency, which is why MultinomialNB slightly outperformed it.

3. Why Did GaussianNB Perform the Worst?
* ❌ Not suited for text classification – GaussianNB assumes continuous numeric features (like heights, weights, etc.), whereas text data is categorical.
* ❌ Less effective with word frequency – Since word frequencies drive sentiment, GaussianNB’s assumption of a normal distribution does not align well with text data.

# Key Takeaways
✅ MultinomialNB is the best choice for sentiment analysis since it effectively models word frequency.

✅ BernoulliNB is also effective, especially when using binary presence/absence features.

✅ GaussianNB is unsuitable for text data and performs worse than both alternatives.